### PR TITLE
[#185] Implement trap region behavior

### LINKED ISSUE #185: Scene Region Trap
Scene regions allow for lots of interesting abilities but they can be overwhelming. A trap region with simple input fields would be best. The region needs:

- [ ] A field for a saving throw DC
- [ ] A field for a saving throw type
- [ ] A field for a damage roll
- [ ] A field for damage type.
- [ ] A field for a chat description

### PR DESCRIPTION
**Pull request by krbz999**
Closes #185.

### CODE REVIEW

File: languages/en.json [hunk 1]
```diff
@@ -132,6 +132,45 @@
   "MYTHACRI.MayhemUpdated": "Updated your mayhem points to ({points}).",
   "MYTHACRI.Pips": "Pips",
   "MYTHACRI.Refresh": "Refresh",
+  "MYTHACRI.REGION": {
+    "TRAP": {
+      "FIELDS": {
```
Comment (krbz999) [inline languages/en.json:137]: Not used directly in the module, this is a native feature of Foundry v12, localizing any data fields automatically.

File: scripts/modules/data/models/behaviors.mjs [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+class TrapData extends foundry.data.regionBehaviors.RegionBehaviorType {
+  /** @override */
+  static events = {
+    [CONST.REGION_EVENTS.TOKEN_ENTER]: foundry.utils.debounce(TrapData.#onTokenMoveIn, 150)
+  };
+
+  /* -------------------------------------------------- */
+
+  /** @override */
+  static LOCALIZATION_PREFIXES = ["MYTHACRI.REGION.TRAP"];
```
Comment (krbz999) [inline scripts/modules/data/models/behaviors.mjs:25]: Regarding localization, this is the array that tells Foundry where to find the strings.

File: scripts/modules/data/models/behaviors.mjs [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+  /**
+   * The script that is executed when a token moves into the region.
+   * @param {RegionEvent} event     The region event data from triggering this behavior.
+   */
+  static async #onTokenMoveIn(event) {
+    const {data: {token}, user} = event;
+
+    const gm = game.users.activeGM;
+    if (![gm, user].includes(game.user)) return;
```
Comment (krbz999) [inline scripts/modules/data/models/behaviors.mjs:63]: Some functions can only be performed by a GM, and some functions should preferably be performed by the user who initiated the movement. All other users are ignored.

File: scripts/modules/data/models/behaviors.mjs [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    if (!this.damage.formula) {
+      ui.notifications.error("MYTHACRI.REGION.TRAP.WARNINGS.FormulaMissing", {localize: true});
+      return;
+    }
+
+    // GM only functions.
+    if (game.user === gm) {
+      if (this.config.toggle) await this.behavior.update({disabled: true});
```
Comment (krbz999) [inline scripts/modules/data/models/behaviors.mjs:78]: If this behavior should be performed once only, this gets allocated to the GM to disable it.

File: scripts/modules/data/models/behaviors.mjs [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+          if (actor) {
+            targets.push({img: token.texture.src, name: token.name, uuid: actor.uuid});
+          }
+        }
+      }
+
+      await roll.toMessage({
+        flavor: await TextEditor.enrichHTML(html, {rollData: rollData, relativeTo: actor}),
+        speaker: gm ? {alias: gm.name} : undefined,
+        "flags.dnd5e.targets": targets,
```
Comment (krbz999) [inline scripts/modules/data/models/behaviors.mjs:114]: The 'targets' are those actors that have a token contained within the region at time of execution.

**Review by Mythacri — COMMENTED** [review]
After a decent amount of testing (and an embarrassing amount of time spent with the region disabled...because I am a fool) the trap region works just as it should! The next dungeon crawl gonna go hard

**Pull request merged**